### ISSUE TITLE
Cannot pass strings with spaces to JVM arguments

### ISSUE THREAD

**Issue by markwoon** [tofi86/universalJavaApplicationStub#14]
I've tried all manner of escapes in an effort to use spaces in parameters passed to JVM Options and nothing seems to work.

Example:

```
<key>JVMOptions</key>
<array>
<string>-Dcom.apple.mrj.application.apple.menu.about.name=&quot;Hello World&quot;</string>
</array>
```

Which just leads to `Error: Could not find or load main class Hello`.

Everything works as expected if I just remove or replace the space between Hello and World.

**Comment by tofi86** [comment]
yep, you're right, that's not working at the moment. sorry.

however, for this specific option, you can set
```
<key>CFBundleName</key>
<string>Hello World</string>
```
in your `Info.plist` file.

**Comment by tofi86** [comment]
Probably related to changes made in 9587c16 following #1 
Will investigate...

**Comment by Arthur2e5** [comment]
```bash
	exec "$JAVACMD" \
			-cp "${JVMClassPath}" \
			-splash:"${ResourcesFolder}/${JVMSplashFile}" \
			-Xdock:icon="${ResourcesFolder}/${CFBundleIconFile}" \
			-Xdock:name="${CFBundleName}" \
			${JVMOptions:+$JVMOptions }\
			${JVMDefaultOptions:+$JVMDefaultOptions }\
			${JVMMainClass}\
			${JVMArguments:+ $JVMArguments}
```

This sloppy quoting breaks things. Consider switching to `#!/bin/bash` and using arrays.

**Comment by tofi86** [comment]
@Arthur2e5 you're right with `#!/bin/bash`, thanks for the hint!

However, I need to digg deeper in the quote escaping issue with your example function, so give me some time for this...
Pull requests are welcome at any time as well ;-)

**Comment by Arthur2e5** [comment]
Yeah, I am not actually sure about how the stub actually parses `JVMOptions`. My current guess is something like string joining (with `"<array elem>"`?) and `system`, which would be `eval` in shell script…

I don't have access to any OS X systems unfortunately, and I am currently too lazy to pick up anything like Hopper. May I ask @markwoon to actually try it with Oracle's Stub, with and w/o `"`?

**Comment by karlvr** [comment]
This issue also occurs if the application bundle lives in a folder containing a space in it. This occurs for me when trying to launch from the disk image, which contains a space in the name.

I'm investigating a solution.

**Comment by karlvr** [comment]
I've confirmed that this is exactly the same problem; just the `JVMOptions`, which are problematic as they contain an `$APP_PACKAGE` replacement. It appears that the expected ways of escaping those isn't working for me. I've tried `\`ing the spaces, and sticking the `$APP_PACKAGE` in ""s...

**Comment by tofi86** [comment]
@Azhrei came up with a quick 'n dirty solution for this in #40:

> I have a Plist file in which the $JAVAROOT value will contain spaces. Unfortunately, when the properties are eval'd to replace variable values, they don't have the correct quotes added around them so they'll work on the eventual command line.
> 
> Thinking about this, I decided adding double quotes was the wrong solution anyway (well, thinking about it AND doing some testing). What if the user put double quotes in their pathname to the app? Yeah, I know they shouldn't, but USERS don't know they shouldn't. (Sigh.)
> 
> So I came up with the following snippet for the JVMOptions variable, around line 171 in the latest version.
> ``` JVMOptions=`$PL -c "print ${JavaKey}:Properties" "${InfoPlistFile}" 2> /dev/null | while read a b c; do [[ "$b" == '=' ]] && echo -n "-D$a=$(eval printf "%q" "$c") " ; done` ```
> 
> Since that code contains the eval, be sure to comment out the following line that does eval echo ... (it's no longer needed).

Closed #40 as duplicate in favor of this thread but cross-posted his solution for easier handling over here...

Sorry guys for not finding a good solution to solve this with 2.0 lately. Will take another attempt soon...

**Comment by Azhrei** [comment]
Well, the better solution is to define shell functions to accomplish some of this so that the code isn't inlined, thus making the script easier to read and modify, but also allowing individual functions to be tested separately.  I'll do that at some point.  In fact, my immediate deadline is past, so I might work on this more today and tomorrow...

What I appreciate most, @tofi86, is the work you did to ferret out the issues involved in making this work in the first place!  That's been my stumbling block for a long time.  Now it's just a matter of implementing the solution in the best possible way, and that kind of coding is my forté. 😄 

What are your feelings about updating working code to use newer idioms?  Like changing out backtics to use $() instead?  If I make those changes and put them in a pull request, would that be acceptable?

Thanks again!

**Comment by tofi86** [comment]
> What I appreciate most, @tofi86, is the work you did to ferret out the issues involved in making this work in the first place!

Thanks! 😊 


> What are your feelings about updating working code to use newer idioms? Like changing out backtics to use $() instead? If I make those changes and put them in a pull request, would that be acceptable?

Yeah, sure, sounds very good! We should use whatever bash has to offer in the meantime. However, we should keep an eye on older Mac OS X versions as we probably have users running Java 6 apps on older systems. So we should check for Bash compatibility up to - let's say - OS X 10.6 (Snow Leopard)...

**Comment by Azhrei** [comment]
Ah, good point.  Stuff like $() is POSIX (and has been for a long time) so I can't imagine a problem there, but the [[ ]] stuff wasn't POSIX, then it was, and then it wasn't again.  Probably best not to change one without plenty of testing. :)

I'll see about creating a VM with Snow Leopard to test against, since I don't have any machines in my office that old that are still running...

**Comment by rambaut** [comment]
I see this issue is still open. I think I am having a similar issue.

I wish to include the $JAVAROOT as the java.library.path property (I have a few jnilibs in there). 

JarBundler creates the following -

```
      <key>Properties</key>
      <dict>
        <key>java.library.path</key>
        <string>$JAVAROOT</string>
        <key>apple.laf.useScreenMenuBar</key>
        <string>true</string>
      </dict>
    </dict>
```

The problem is that $JAVAROOT is an absolute path and contains spaces. This causes the error:
`Could not find or load main class v1.10pre20170801.BEAUti`

So the space is breaking the path and it is being interpreted as the main class.

**Comment by rambaut** [comment]
Hi, I see this issue is still open and it is causing me a few problems. I wish to include JAVAROOT as the java.library.path property (I have some JNI libs there). 

JarBundler creates:
```
      <dict>
        <key>java.library.path</key>
        <string>$JAVAROOT</string>
        <key>apple.laf.useScreenMenuBar</key>
        <string>true</string>
      </dict>
```

Which then comes up against the issue with the spaces in the path.

If I edit universalJavaApplicationStub to have:

```
	exec "$JAVACMD" \
			-cp "${JVMClassPath}" \
			-splash:"${ResourcesFolder}/${JVMSplashFile}" \
			-Xdock:icon="${ResourcesFolder}/${CFBundleIconFile}" \
			-Xdock:name="${CFBundleName}" \
			${JVMOptions:+$JVMOptions }\
			-Djava.library.path="${JAVAROOT}" \
			${JVMDefaultOptions:+$JVMDefaultOptions }\
			${JVMMainClass}\
			${JVMArguments:+ $JVMArguments}\
```
it works fine.

Note the explicit `-Djava.library.path` with `${JAVAROOT}` in quotes. 

As the postings above suggest, there doesn't seem to be away to get these quotes to come through from the Info.plist.

**Comment by Azhrei** [comment]
That works for you, but if the ```JAVAROOT``` variable is empty, it will fail (for everyone else that doesn't use the variable).  However, if you replace that line with this it will work for everyone:

```${JAVAROOT+"-Djava.library.path=$JAVAROOT"}```

That variable substitution uses something called a _parameter expansion modifier_ (technically it should be a _parameter **substitution** modifier_, but the legacy term is typically used).  It says that if the ```JAVAROOT``` variable has something stored in it, do a substitution but don't use the variable's value; instead, use the contents of the double-quoted string that follows.  And that string contains the command line option you need.  Since it's inside quotes, the spaces inside the variable will be preserved.

I will make this change and it will probably get into the imminent 1.4.1.8 build.  Thanks for the suggestion!

**Comment by rambaut** [comment]
Thanks. I wasn't suggesting I had a general solution, just one I could use in my application. But yours is much better. I will try it out.

**Comment by Azhrei** [comment]
This is true.

My focus was not on solving the Java problem, only on ensuring that embedded spaces were being preserved by the shell.  I don't wish to misrepresent that what I suggested was going to solve all problems for everyone, only that it was a sample for how to properly quote an environment variable that might contain spaces.

For the Java problem, there is no builtin solution (apparently).  Each platform defaults to a potentially different list of directories regarding `java.library.path` and there doesn't seem to be any way to append/prepend to that list from the command line.  Being able to both append and prepend to various builtin search paths causes a truly bloated command line option list (q.v. GCC C compiler in regards to include directories).

Oh well. :)

**Comment by tofi86** [comment]
I'll try to have a look at this in the next couple of weeks. Thanks for reporting this issue again.

**Comment by Dylan-M** [comment]
@tofi86 Where are we on this issue? My builds are still using 1.0.1 and I'm looking to get updated code, but don't want to still have this issue.

**Comment by tofi86** [comment]
I think I did some testing for this a couple of weeks ago but couldn't find a solution for the problem.

With version 2 I switched to Bash instead of Sh and with the latest commits to `develop` there have been a couple of fixes for wrong quoting.

However, I just couldn't find a solution for this one here...

Since I've been working a lot more with Bash arrays lately I could probably give it another try with Bash arrays in the next couple of weeks...

Help is always welcome! :)

@Dylan-M you're still on 1.0.1 - why? did it work back then?

**Comment by Azhrei** [comment]
I don't think it'll be portable, but to be honest...

The best way to handle this is likely to build the command line and feed it as stdin to an interactive shell.  This avoids the problems inherent with quoting certain characters, but not all platforms can handle this (i.e. Windows, although PowerShell probably can).  Another option would be to create a file and then execute the file, but then you have these "deposits" on your hard drive that someone has to clean up at some point...

**Comment by Dylan-M** [comment]
@tofi86 No, still on 1.0.1 because I took a hiatus from the project and the other devs don't understand this script 💭

**Comment by tofi86** [comment]
Update: I implemented an improved handling of MainClass arguments with spaces in #57 (arguments to the `*.app`) and #58 (arguments via `Info.plist`).

The new behaviour can be tested on the `develop` branch. Feedback is kindly appreciated (please use issue #57 or #58 for providing feedback!).

I will try to implement this as well for the `JVMOptions` mentioned by issue author @markwoon. Stay tuned.

**Comment by tofi86** [comment]
Alright, this should also be working now for `JVMOptions` Plist key.

Would love to recieve feedback from you guys! Check out the latest commits from `develop` branch and give it a try!

**Comment by tofi86** [comment]
This is the last issue blocking the 3.0 release at the moment.

I hope that I wasn't too tired and that I really fixed the bug correctly... 😴  However, I am very confident! 👍

**Comment by tofi86** [comment]
Alright, I have been testing this a second time in dev and will now release the fix with version 3.0.0.
